src/odemis/driver/andorshrk.py

```diff
@@ -2072,6 +2156,115 @@ def scan():
 
+class ShamrockBus(model.HwComponent):
+    """
+    Create several Shamrock children, sharing the same library instance
+    Used to instantiate multiple Andor spectrographs, as they cannot be instantiated separately.
+    """
+    def __init__(self, name: str, role: str, children: Dict[str, Dict], daemon=None, **kwargs):
+        """
+        :param name: name of the component
+        :param role: role of the component (typically not useful)
+        :param children: abritrary role -> arguments for Shamrock. The arguments must contain a "device"
+        argument with the serial number of the spectrograph (not just a number). "fake" is also possible,
+        in which case a simulator is used.
+        """
+        super().__init__(name, role, daemon=daemon, **kwargs)
+
+        # prepare the children definitions, by matching the serial numbers from "device" argument
+        # The simulated devices are separated, as they don't need the library
+        shamrocks = {}
+        simulated = []
+        for ckwargs in children.values():
+            try:
+                # device will be replaced by the device number once we find it
+                device = ckwargs.pop("device")
+            except KeyError:
+                raise ValueError(f"Missing 'device' argument in child component {kwargs['name']}")
+            if not isinstance(device, str):
+                raise ValueError(f"The 'device' argument should be a string, but got \"{device}\"")
+
+            if device == "fake":
+                simulated.append(ckwargs)
+            else:
+                shamrocks[device] = ckwargs
+
+        for ckwargs in simulated:
+            dev = Shamrock(device="fake", daemon=daemon, **ckwargs)
+            self.children.value.add(dev)
+
+        if shamrocks:
+            self._dll = ShamrockDLL()
+            self.Initialize()
+
+            try:
+                # scan the devices
+                sn_c = create_string_buffer(64)
+                for n in range(self.GetNumberDevices()):
+                    self._dll.ShamrockGetSerialNumber(n, sn_c)
+                    sn = sn_c.value.decode('latin1')
+                    if sn in shamrocks:
+                        # create the child
+                        logging.debug("Found matching Andor Shamrock with S/N %s", sn)
+                        ckwargs = shamrocks.pop(sn)
+                        dev = Shamrock(device=n, daemon=daemon, **ckwargs)
+                        self.children.value.add(dev)
+                    else:
+                        logging.info("Skipping Andor Shamrock with S/N %s", sn)
+
+                    # In (the unlikely) case we have found all the spectrographs we need, but there are
+                    # more spectrographs, stop early.
+                    if not shamrocks:
+                        break
+            except Exception:
+                # One of the child init went wrong, just clean up and pass on the bad news
+                self.Close()
```

Comment:
```suggestion
                self.terminate()
```
Shouldn't you also terminate the appended children as a part of the clean-up?